### DISCUSSION TITLE
Weird behavior from "indepth analysis language" and "recently used languages"

### DISCUSSION THREAD

**Discussion by stefanusgusega** [lowlighter/metrics#971] [Q&A]
![image](https://user-images.githubusercontent.com/52844951/160251612-5bc36824-90df-4d8d-902c-c0af07046213.png)
I just implemented this feature, and wondering why did my recently used languages not show any language and the percentage. Also, on in depth analysis of languages, why did the stats says *0 lines* but there are percentage values. How do I fix this? Thank you.

**Comment by lowlighter** [comment]
Hi 👋 !

Could you run your workflow with `debug: yes` and post the logs here ?
I think the `indepth` mode just failed in your case for whatever reason, but with just the resulting output it's hard to tell why

___

> Also, on in depth analysis of languages, why did the stats says 0 lines but there are percentage values?

It's because it make a raw estimation first from GitHub GraphQL API first. 

Lines and accurate stats are computed by cloning you repos on the runner and iterating directly on the git logs of it, which is quite expensive in term of resources, but helps to provide more accurate stats along with new ones like the number of lines

**Reply by stefanusgusega** [reply]
Allright, I've just added the ```debug: yes```, and I will post the logs here later.

> Lines and accurate stats are computed by cloning you repos on the runner and iterating directly on the git logs of it, which is quite expensive in term of resources, but helps to provide more accurate stats along with new ones like the number of lines

Is it the expected behavior? So, I can't display the lines for now? Or, are there some workarounds for this issue? 

Also, in the habit chart section, the languages just disappeared. Is this caused by indepth analysis too?
![image](https://user-images.githubusercontent.com/52844951/160314089-10c8d009-e5a4-49be-88a9-7bc90fa9a8af.png)

Thank you in advance!

**Reply by stefanusgusega** [reply]
For the logs, can you access [this](https://github.com/stefanusgusega/stefanusgusega/runs/5713520936?check_suite_focus=true) instead of copying all raw logs here? Thank you in advance.

**Reply by lowlighter** [reply]
> Allright, I've just added the `debug: yes`, and I will post the logs here later.

Thanks a lot 👍 

> Is it the expected behavior? So, I can't display the lines for now? Or, are there some workarounds for this issue?

Lines are exclusive to `indepth` mode and there's no workaround for it. It's because GitHub API does not provide such detailed information about lines per user per language per repo.

The indepth mode actually clones your repo on the GitHub runner and use `git blame`/`git log` to filter out only the lines tagged with your username (or what is provided in `commit_authoring` option) and map them to the file language found by [GitHub linguist](https://github.com/github/linguist). This is why it also have access to number of lines. (You can read more about this mode [here](https://github.com/lowlighter/metrics/blob/master/source/plugins/languages/README.md#-indepth-mode))

The default algorithm (which most of apps similar to metrics also use) is just cumulative stats from what is displayed by GitHub on the repository side-bar (but GitHub doesn't make any difference about the author), so it is less representative if you're not the sole contributor of a given repository

> Also, in the habit chart section, the languages just disappeared. Is this caused by indepth analysis too?

Yes, actually both charts are the same (if indepth mode is disabled the habits language chart is "hidden") so if one fails, the other one will probably fails too.

The "recently used language" chart used to be in `habits` plugin rather than `languages` plugin. This is mostly due to legacy reason

**Reply by lowlighter** [reply]
> For the logs, can you access [this](https://github.com/stefanusgusega/stefanusgusega/runs/5713520936?check_suite_focus=true) instead of copying all raw logs here? Thank you in advance.

Yes, should be fine 👍

**Reply by lowlighter** [reply]
For most used section, I think it's probably because you reached the maximum timeout (default is 15min but you can increase it to up 30min using [`plugin_languages_analysis_timeout`](https://github.com/lowlighter/metrics/blob/master/source/plugins/languages/README.md#plugin_languages_analysis_timeout) option):
```
metrics/compute/stefanusgusega/plugins > languages > Reached maximum execution time of 15m for analysis
```

___

For recently used section, it's because your [`commits_authoring`](https://github.com/lowlighter/metrics/tree/a0eb7985243b8657d9d2cb19ddb65afc4bcb1c9d/source/plugins/base#commits_authoring) option is not properly set. By default, it uses your username (stefanusgusega) but as you can see from this example [commit](https://api.github.com/repos/stefanusgusega/tl-covid-cough-classification/commits/bce4ad02e26bab29a37c720f8a8c0e0b069f580c) your commits user/mail does not match your GitHub login, which is why they're getting skipped.

You could just be more permissive with the regexp matcher:
```yml
commits_authoring: stefanus
```

Or use a more detailed version, if you want to be sure to not have any collision with other people with the same name:
```yml
commits_authoring: stefanusgusega, stefanus gusega gunawan
```

**Reply by stefanusgusega** [reply]
> For most used section, I think it's probably because you reached the maximum timeout (default is 15min but you can increase it to up 30min using [plugin_languages_analysis_timeout](https://github.com/lowlighter/metrics/blob/master/source/plugins/languages/README.md#plugin_languages_analysis_timeout) option)

If I set to the 30 mins, will the *line* information appeared? What if I reached the maximum timeout again even if I already set it on 30 mins?

> it's because your [commits_authoring](https://github.com/lowlighter/metrics/tree/a0eb7985243b8657d9d2cb19ddb65afc4bcb1c9d/source/plugins/base#commits_authoring) option is not properly set.

I've noticed this a few hours ago, thank you for clarifying this. Now, the runner is running with configured ```commits.authoring``` set to my username, my email, and my university email, and let's see if it will be fixed.

Thank you in advance!

**Reply by stefanusgusega** [reply]
The *recent language* section seems fixed. And, I have 2 questions.

1. As stated in this debug log, does this plugin skip private repositories? Because the repository that I quoted is exactly a private repository.
> metrics/compute/stefanusgusega/plugins > languages > indepth > an error occured while processing stefanusgusega/KaryakuOrMaknart, skipping...

2. Does this plugin only scan repository that I created? Or, it considers my commit to other public repository? Because, I noticed that I think the debug log states only repository that I created. 

Thank you in advance.

**Reply by lowlighter** [reply]
Sorry for the late reply

> 1. As stated in this debug log, does this plugin skip private repositories? Because the repository that I quoted is exactly a private repository.

It isn't supposed to skip private repository (unless the token used doesn't have access to it), but if there is any error during the processing it'll be skipped (like it seems to be in this case) to avoid making the whole indepth analysis failed. Also to avoid "code leaks", the action isn't very verbose on this kind of error so it's kind of hard to debug and cannot take a look at it myself

> 2. Does this plugin only scan repository that I created? Or, it considers my commit to other public repository? Because, I noticed that I think the debug log states only repository that I created.

If you have not explicitely configured [`repositories_affiliations`](https://github.com/lowlighter/metrics/blob/master/source/plugins/base/README.md#repositories_affiliations) to another constraint, yes it default to `owner` only, but you can extend it to `collaborator`, `organization_member` or the broadest one which is no constraints: `""`

Additionally, [`repositories_forks`](https://github.com/lowlighter/metrics/blob/master/source/plugins/base/README.md#repositories_forks) are also ignored by default in case you were wondering, but can be opted-in too

Hope it helps 👍

**Reply by stefanusgusega** [reply]
> It isn't supposed to skip private repository (unless the token used doesn't have access to it)

The token set for this metrics is stated below. Is this true? Or am I missed any permissions?
![image](https://user-images.githubusercontent.com/52844951/161468756-4b8c4125-45ba-40e1-a388-c04ad83fc655.png)

> If you have not explicitely configured [repositories_affiliations](https://github.com/lowlighter/metrics/blob/master/source/plugins/base/README.md#repositories_affiliations) to another constraint, yes it default to owner only, but you can extend it to collaborator, organization_member or the broadest one which is no constraints: ""

Thank you for this clarification, maybe you should state this in this repository? Or is it explained before?

**Reply by lowlighter** [reply]
> The token set for this metrics is stated below. Is this true? Or am I missed any permissions? 

No `repo` is basically for everything that is private so it's sufficient permissions.
The languages plugin will be reworked soon to support additional features so it may be fixed without anything to do from your side in the future

> Thank you for this clarification, maybe you should state this in this repository? Or is it explained before?

It's only specified in the plugin documentation, but I don't think it's mentioned in the "Getting started" guides.
Maybe it'd be a good idea to at least link `core`/base` plugins documentation in them
